// @anna-ai/cli/dist/publish-BYWuujP3.js
import "./credentials-BTv2IfUZ.js";
import "./apps-CCdtLmxQ.js";
import { CliError } from "./client-D-_z1ALk.js";
import "./bundled-executas-B6b8gIfp.js";
import "./executas-CK3er6f9.js";
import "./executa-cache-CXiEgFZY.js";
import { runExecutaPublish } from "./executa-publish-CkPAB34b.js";
import "./manifest-Bljz8Y6T.js";
import { withErrorHandling } from "./_lifecycle-shared-BpSOfVCP.js";
import "./app-cache-Bl7cE5fm.js";
import "./app-bundle-upload-BhAYo6yj.js";
import { runAppsPublish } from "./apps-publish-Dgi4lBlu.js";
import { resolve } from "node:path";
import { existsSync } from "node:fs";
import { red, yellow } from "kleur/colors";

//#region src/publish/detector.ts
function detectProjectKind(cwd) {
	const hasManifest = existsSync(resolve(cwd, "manifest.json"));
	const hasExecuta = existsSync(resolve(cwd, "executa.json"));
	if (hasManifest && hasExecuta) return "ambiguous";
	if (hasManifest) return "app";
	if (hasExecuta) return "executa";
	return "unknown";
}

//#endregion
//#region src/commands/publish.ts
async function runTopLevelPublish(opts) {
	return withErrorHandling(async () => {
		const cwd = resolve(opts.cwd ?? process.cwd());
		const kind = detectProjectKind(cwd);
		if (kind === "ambiguous") throw new CliError(`cannot auto-detect: ${cwd} has BOTH manifest.json and executa.json. Run \`anna-app apps publish\` or \`anna-app executa publish\` explicitly.`, 4);
		if (kind === "unknown") throw new CliError(`no manifest.json or executa.json found in ${cwd}`, 4);
		if (kind === "app") return runAppsPublish({
			cwd,
			bump: opts.bump,
			noWrite: opts.noWrite,
			dryRun: opts.dryRun,
			account: opts.account,
			json: opts.json,
			noBundle: opts.noBundle,
			bundleDir: opts.bundleDir
		});
		return runExecutaPublish({
			cwd,
			bump: opts.bump,
			noWrite: opts.noWrite,
			publish: opts.publish,
			dryRun: opts.dryRun,
			account: opts.account,
			json: opts.json
		});
	});
}

//#endregion
export { runTopLevelPublish };
