// @anna-ai/cli/dist/sampling-BcML4teS.js
import { canonicalHost, getAccount } from "./credentials-BTv2IfUZ.js";
import { resolve } from "node:path";
import { existsSync, readFileSync } from "node:fs";

//#region src/executa/sampling.ts
var SamplingBridge = class {
	mocks = [];
	cachedSession = null;
	constructor(opts) {
		this.opts = opts;
		if (opts.mode === "mock" && opts.mockFile) {
			const path = resolve(opts.mockFile);
			if (existsSync(path)) for (const line of readFileSync(path, "utf8").split(/\r?\n/)) {
				if (!line.trim() || line.startsWith("#")) continue;
				try {
					this.mocks.push(JSON.parse(line));
				} catch {}
			}
		}
	}
	async createMessage(req) {
		if (this.opts.mode === "off") throw withCode(new Error("sampling disabled — rerun without `--no-sampling`"), -32008);
		if (this.opts.mode === "mock") return this.mockMessage(req);
		return this.realMessage(req);
	}
	mockMessage(req) {
		const content = collectContent(req);
		const entry = this.mocks.find((m) => m.ns === "sampling" && m.method === "createMessage" || m.ns === "llm" && m.method === "complete") ?? null;
		const matched = entry && (!entry.match?.contentIncludes || content.includes(entry.match.contentIncludes)) ? entry : this.mocks.find((m) => m.ns === "sampling" && m.method === "createMessage" || m.ns === "llm" && m.method === "complete");
		if (matched && matched.result) return normaliseSamplingResult(matched.result, "mock-model");
		return {
			role: "assistant",
			content: {
				type: "text",
				text: "(mock) no fixture matched"
			},
			model: "mock-model",
			stopReason: "endTurn"
		};
	}
	account() {
		const acc = getAccount(this.opts.account);
		if (!acc) throw withCode(new Error("no PAT on disk — run `anna-app login --host <nexus-url>` first (or pass `--mock-sampling <fixture>` / `--no-sampling`)"), -32001);
		if (acc.expires_at && acc.expires_at < Math.floor(Date.now() / 1e3)) throw withCode(new Error("PAT expired — run `anna-app login` again"), -32001);
		return acc;
	}
	async mint() {
		if (this.cachedSession && this.cachedSession.expiresAt - 30 > Math.floor(Date.now() / 1e3)) return this.cachedSession;
		const acc = this.account();
		if (!this.opts.appSlug) throw withCode(new Error("sampling bridge has no app_slug — pass `--app-slug <slug>` to `anna-app executa dev` so nexus can attribute the spend"), -32602);
		const url = `${canonicalHost(acc.host)}/api/v1/anna-apps/dev/session/mint`;
		const res = await fetch(url, {
			method: "POST",
			headers: { "content-type": "application/json" },
			body: JSON.stringify({
				pat: acc.pat,
				kind: "complete",
				app_slug: this.opts.appSlug
			})
		});
		if (!res.ok) {
			const text = await res.text().catch(() => "");
			throw withCode(new Error(`session.mint failed: HTTP ${res.status} ${text}`), -32e3);
		}
		const body = await res.json();
		this.cachedSession = {
			appSessionToken: body.app_session_token,
			expiresAt: Math.floor(Date.now() / 1e3) + (body.expires_in || 600)
		};
		return this.cachedSession;
	}
	async realMessage(req) {
		const acc = this.account();
		const session = await this.mint();
		const body = mcpToCompleteBody(req);
		const res = await fetch(`${canonicalHost(acc.host)}/api/v1/copilot/app/complete`, {
			method: "POST",
			headers: {
				"content-type": "application/json",
				authorization: `Bearer ${session.appSessionToken}`
			},
			body: JSON.stringify(body)
		});
		if (!res.ok) {
			const text = await res.text().catch(() => "");
			throw withCode(new Error(`HTTP ${res.status}: ${text}`), -32e3);
		}
		const out = await res.json();
		return normaliseSamplingResult(out, out.model ?? "unknown");
	}
};
function withCode(err, code) {
	err.rpcCode = code;
	return err;
}
/** Collect a single string view of a sampling request — used for mock matching. */
function collectContent(req) {
	const parts = [];
	if (req.systemPrompt) parts.push(req.systemPrompt);
	for (const m of req.messages ?? []) {
		const c = m.content;
		if (typeof c === "string") parts.push(c);
		else if (c && typeof c === "object" && "text" in c) parts.push(c.text);
	}
	return parts.join("\n");
}
/**
* Translate an MCP `sampling/createMessage` request to the nexus
* `/api/v1/copilot/app/complete` body. Mapping is intentionally lossy —
* we forward only the fields nexus understands.
*/
function mcpToCompleteBody(req) {
	const messages = (req.messages ?? []).map((m) => ({
		role: m.role,
		content: typeof m.content === "string" ? m.content : m.content.text
	}));
	const out = { messages };
	if (req.systemPrompt) out.system = req.systemPrompt;
	if (req.maxTokens != null) out.max_tokens = req.maxTokens;
	if (req.temperature != null) out.temperature = req.temperature;
	if (req.stopSequences) out.stop = req.stopSequences;
	const hint = req.modelPreferences?.hints?.[0]?.name;
	if (hint) out.model_hint = hint;
	if (req.metadata) out.metadata = req.metadata;
	return out;
}
/** Coerce arbitrary nexus / fixture output into a `SamplingResult`. */
function normaliseSamplingResult(raw, fallbackModel) {
	if (!raw || typeof raw !== "object") return {
		role: "assistant",
		content: {
			type: "text",
			text: String(raw ?? "")
		},
		model: fallbackModel,
		stopReason: "endTurn"
	};
	const r = raw;
	if (r.role === "assistant" && r.content && typeof r.content === "object") return r;
	const text = typeof r.text === "string" && r.text || (typeof r.content === "string" ? r.content : "") || (typeof r.message === "string" ? r.message : "") || "";
	return {
		role: "assistant",
		content: {
			type: "text",
			text
		},
		model: r.model ?? fallbackModel,
		stopReason: r.stop_reason ?? r.stopReason ?? "endTurn",
		usage: r.usage ?? void 0,
		_meta: { source: "nexus" }
	};
}

//#endregion
export { SamplingBridge };
